src/runtime/crdt/crdt-singleton.ts

```diff
@@ -26,3 +44,68 @@ type SingletonChange<T> = CRDTChange<CRDTSingletonTypeRecord<T>>;
 
 type SingletonModel<T> = CRDTModel<CRDTSingletonTypeRecord<T>>;
 
+export class CRDTSingleton<T> implements SingletonModel<T> {
+  private collection = new CRDTCollection<T>();
+
+  merge(other: SingletonData<T>):
+      {modelChange: SingletonChange<T>, otherChange: SingletonChange<T>} {
+    this.collection.merge(other);
+    // We cannot pass through the collection ops, so always return the updated model.
+    const change: SingletonChange<T> = {
+      changeType: ChangeType.Model,
+      modelPostChange: this.collection.getData()
+    };
+    return {modelChange: change, otherChange: change};
+  }
+
+  applyOperation(op: SingletonOperation<T>): boolean {
+    if (op.type === SingletonOpTypes.Clear) {
+      return this.clear(op.actor, op.clock);
+    }
+    if (op.type === SingletonOpTypes.Set) {
+      // Remove does not require an increment, but the caller of this method will have incremented
+      // its version, so we hack a version with t-1 for this actor.
+      const removeClock =
+          (new Map(op.clock)).set(op.actor, op.clock.get(op.actor) - 1);
+      if (!this.clear(op.actor, removeClock)) {
+        return false;
+      }
+      const addOp: CollectionOperation<T> = {
+        type: CollectionOpTypes.Add,
+        added: op.value,
+        actor: op.actor,
+        clock: op.clock,
+      };
+      if (!this.collection.applyOperation(addOp)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  getData(): SingletonData<T> {
+    return this.collection.getData();
+  }
+
+  getParticleView(): RawSingleton<T>|null {
+    // Return any value.
+    return [...this.collection.getParticleView()].sort()[0] || null;
+  }
+
+  private clear(actor: string, clock: VersionMap): boolean {
+    // Clear all existing values if our clock allows it.
+    for (const value of this.collection.getData().values.keys()) {
+      const removeOp: CollectionOperation<T> = {
+        type: CollectionOpTypes.Remove,
+        removed: value,
+        actor,
+        clock,
+      };
+      // If any value fails to remove, we haven't cleared the value and we fail the whole op.
```

Comment:
I think this is no longer true given the changes in remove for collections - however, we also don't want it to be true. Clear should only remove values seen by the generator of the clear op; this means that our "singleton" can sometimes be multi-valued.

Reply:
are you saying this should just be:
this.collection.applyOperation(removeOp);
without checking the return value? ie we are fine if some values are still there.
or should we return false if an op fails, so the caller can sync the model if necessary?

Reply:
based on our chat, changing this to ignore the return value.
modified to test to explain this behaviour (two actors can set two different values simultaneously)